File: climate.py
import csv
import time
from datetime import datetime 
from collections import defaultdict

#climate data class to hold record in class instance
class ClimateData:
    def __init__(self, local_date, speed_max_gust, total_precipitation, min_temperature, max_temperature, mean_temperature):
        self.local_date = local_date
        self.speed_max_gust = speed_max_gust
        self.total_precipitation = total_precipitation
        self.min_temperature = min_temperature
        self.max_temperature = max_temperature
        self.mean_temperature = mean_temperature

#function to process CSV input file
def load_climatedata(filename):
    datalist = []

    with open(filename) as numbers:
        numberdata = csv.reader(numbers, delimiter=',')
        next(numberdata)
        for row in numberdata:
            try:
                local_date = (row[0])
                speed_max_gust = int(row[1])
                total_precipitation = float(row[2])
                min_temperature = float(row[3])
                max_temperature = float(row[4])
                mean_temperature = float(row[5])
                climate = ClimateData(local_date, speed_max_gust,total_precipitation,min_temperature,max_temperature,mean_temperature)
                datalist.append(climate)
            except ValueError:
                continue

    return datalist

#convert date to readable format: YYYY-MM-DD to Month Day Year or Month Year as per the assignment guidelines (February 2010, March 2010 or February dd YYYY etc)
def convert_date(date, format):
    months = {"01": "January", "02": "February", "03": "March", "04": "April", "05": "May", "06": "June", "07": "July", "08": "August", "09": "September", "10": "October", "11": "November", "12": "December"}
    if format == "day":
        return (f"{months[date.split('-')[1]]} {date.split('-')[2].split(' ')[0]} {date.split('-')[0]}")
    elif format == "month":
        return (f"{months[date.split('-')[1]]} {date.split('-')[0]}")
    
# Aux function validate date
def validate_date(date_str):
    try:
        date_object = datetime.strptime(date_str, "%YY-MM-DD")
        return date_object >= datetime(2010, 4, 18) and date_object <= datetime(2024, 2, 8)
    except ValueError:
        return False

# Analyze data - q2. Must include Highest total_precipitation (month and year). Day with highest max gust speed - print full date. Day with largest temperature fluctation
def data_analysis(data: list[ClimateData]):

    max_gust_date = None
    max_gust = 0
    max_precipitation = 0
    percipitation_per_month = defaultdict(float)
    max_temp_fluctuation = 0
    max_temp_fluctuation_date = None

    for item in data:
        month_key = convert_date(item.local_date,"month")
        percipitation_per_month[month_key] += item.total_precipitation 

        if item.max_temperature - item.min_temperature > max_temp_fluctuation:
            max_temp_fluctuation = item.max_temperature - item.min_temperature
            max_temp_fluctuation_date = item.local_date
        
        if item.speed_max_gust > max_gust:
            max_gust = item.speed_max_gust
            max_gust_date = item.local_date
    
    max_monthly_precipitation, max_precipitation = max(percipitation_per_month.items(), key=lambda value: value[1])
    print("\nClimate Data Analysis of Kitchener-Waterloo from 2010 to 2024")
    print(f"The month with the most precipitation (mm): {max_monthly_precipitation} with {max_precipitation:.2f}mm")
    print(f"The date with the highest gust speed (km/h): {convert_date(max_gust_date, 'day')} with {max_gust}km/h ")
    print(f"The date with the highest temperature fluctation (°C): {convert_date(max_temp_fluctuation_date, 'day')} with {max_temp_fluctuation:.2f}°C\n")

    return

#User report - multiple options for user to select from. This will be "main program" and will call functions to generate specific reports
def user_report(data: list[ClimateData]):
    while True:
        print("Choose one of the following options:")
        print("1) Average monthly weather data")
        print("2) Weather records between two dates")
        print("3) Exit")
        command = input("Enter your option number: ")

        if command == '1':
            start_time = time.time()
            avg_monthly_data(data)
            end_time = time.time()
            print(f"The average by month report took {end_time - start_time} seconds to run")
        elif command == '2':
            start_date = input("Enter a start date (YYYY-MM-DD): ")
            while not validate_date(start_date):
                print("Invalid date! Please enter a date between 2010-04-18 and 2024-02-08 in the format YYYY-MM-DD.")
                start_date = input("Enter a start date (YYYY-MM-DD): ")
            end_date = input("Enter a end date (YYYY-MM-DD): ")
            while not validate_date(end_date):
                print("Invalid date! Please enter a date between 2010-04-18 and 2024-02-08 in the format YYYY-MM-DD.")
                end_date = input("Enter a end date (YYYY-MM-DD): ")
            
            date_range_report(data, start_date, end_date)
        elif command == '3':
            break
        else:
            print("Invalid option number. Please try again")

    return
            
            

def avg_monthly_data(data: list[ClimateData]):

    return

def date_range_report(data: list[ClimateData], start_date, end_date):

    return



        
if __name__ == "__main__":
    data = load_climatedata("climate-daily.csv")
    data_analysis(data)

File: climate.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <ctype.h>


typedef struct {
    char date[20];
    float maxGust;
    float totalPrecipitation;
    float minTemperature;
    float maxTemperature;
    float meanTemperature;
} ClimateRecord;

typedef struct {
    char month[8];
    float totalPrecipitation;
} MonthlyPrecipitation;


// Function to parse a line from the CSV file
ClimateRecord* load_climate_data(char* line) {
    ClimateRecord* record = malloc(sizeof(ClimateRecord));
    if (record == NULL) {
        fprintf(stderr, "Failed to allocate memory for record.\n");
        return NULL;
    }

    char* token = strtok(line, ",");
    if (token == NULL || token[0] == '\0') {
        free(record);
        return NULL;
    }
    strcpy(record->date, token);

    token = strtok(NULL, ",");
    if (token == NULL || token[0] == '\0') {
        free(record);
        return NULL;
    }
    record->maxGust = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL || token[0] == '\0') {
        free(record);
        return NULL;
    }
    record->totalPrecipitation = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL || token[0] == '\0') {
        free(record);
        return NULL;
    }
    record->minTemperature = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL || token[0] == '\0') {
        free(record);
        return NULL;
    }
    record->maxTemperature = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL || token[0] == '\0') {
        free(record);
        return NULL;
    }
    record->meanTemperature = atof(token);

    return record;
}

// Function to analyze the data
void analyzeData(ClimateRecord* records[], int record_count) {
    char* maxGustDate = NULL;
    int maxGust = 0;
    char* maxTempFluctuationDate = NULL;
    float maxTempFluctuation = 0;

    int monthlyPrecipitationSize = 100;
    MonthlyPrecipitation* monthlyPrecipitations = malloc(monthlyPrecipitationSize * sizeof(MonthlyPrecipitation));
    int monthlyPrecipitationCount = 0;


    for (int i = 0; i < record_count; i++) {
        ClimateRecord* record = records[i];
        if (record == NULL) {
            continue;
        }

        char month[8];
        strncpy(month, record->date, 7);
        month[7] = '\0';

        int found = 0;
        for (int j = 0; j < monthlyPrecipitationCount; j++) {
            if (strcmp(monthlyPrecipitations[j].month, month) == 0) {
                monthlyPrecipitations[j].totalPrecipitation += record->totalPrecipitation;
                found = 1;
                break;
            }
        }
        
        if (!found) {
            if (monthlyPrecipitationCount == monthlyPrecipitationSize) {
                monthlyPrecipitationSize *= 2;
                monthlyPrecipitations = realloc(monthlyPrecipitations, monthlyPrecipitationSize * sizeof(MonthlyPrecipitation));
            }
            strcpy(monthlyPrecipitations[monthlyPrecipitationCount].month, month);
            monthlyPrecipitations[monthlyPrecipitationCount].totalPrecipitation = record->totalPrecipitation;
            monthlyPrecipitationCount++;
        }

        if (record->maxGust > maxGust) {
            maxGust = record->maxGust;
            maxGustDate = record->date;
        }

        float tempFluctuation = record->maxTemperature - record->minTemperature;
        if (tempFluctuation > maxTempFluctuation) {
            maxTempFluctuation = tempFluctuation;
            maxTempFluctuationDate = record->date;
        }
    }

    char maxPrecipitationMonth[8] = "";
    float maxPrecipitation = 0;
    for (int i = 0; i < monthlyPrecipitationCount; i++) {
        if (monthlyPrecipitations[i].totalPrecipitation > maxPrecipitation) {
            maxPrecipitation = monthlyPrecipitations[i].totalPrecipitation;
            strcpy(maxPrecipitationMonth, monthlyPrecipitations[i].month);
        }
    }

    printf("\nClimate Data Analysis of Kitchener-Waterloo from 2010 to 2024\n");
    printf("The month with the most precipitation (mm): %s with %.2fmm\n", maxPrecipitationMonth, maxPrecipitation);
    printf("The date with the highest gust speed (km/h): %s with %dkm/h\n", maxGustDate, maxGust);
    printf("The date with the highest temperature fluctation (°C): %s with %.2f°C\n", maxTempFluctuationDate, maxTempFluctuation);
}


// Function to print the records between two dates
void date_range_report(ClimateRecord** records, int record_count, char* start_date, char* end_date) {
    for (int i = 0; i < record_count; i++) {
        ClimateRecord* record = records[i];
        if (record == NULL) {
            continue;
        }
        if (strcmp(record->date, start_date) >= 0 && strcmp(record->date, end_date) <= 0) {
            printf("\nDate: %s\n", record->date);
            printf("Max Gust: %.2fkm/h\n", record->maxGust);
            printf("Total Precipitation: %.2fmm\n", record->totalPrecipitation);
            printf("Min Temperature: %.2f°C\n", record->minTemperature);
            printf("Max Temperature: %.2f°C\n", record->maxTemperature);
            printf("Avg Temperature: %.2f°C\n\n", (record->minTemperature + record->maxTemperature) / 2);
        }
    }
}

// Function to generate user reports
void userGeneratedReport(ClimateRecord** records, int record_count) {
    int choice;
    char start_date[20];
    char end_date[20];
    do {
        printf("\nChoose one of the following options:\n");
        printf("1) Average monthly weather data\n");
        printf("2) Weather records between two dates\n");
        printf("3) Exit\n");
        printf("Enter your option number: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:{
                clock_t start_time = clock();
                //avg_monthly_data(records, record_count);
                clock_t end_time = clock();
                double duration = ((double) (end_time - start_time)) / CLOCKS_PER_SEC;
                printf("The average by month report took %f seconds to run\n", duration);
                break;
            }
            case 2:
                printf("Enter the start date (YYYY-MM-DD): ");
                scanf("%s", start_date);
                printf("Enter the end date (YYYY-MM-DD): ");
                scanf("%s", end_date);
                date_range_report(records, record_count, start_date, end_date);
                break;
            case 3:
                break;
            default:
                printf("Invalid choice. Please try again.\n");
        }
    } while (choice != 3);
}

// main function to run climate data program
int main() {
    int record_count = 0;
    int record_capacity = 200;

    // Allocate initial memory
    ClimateRecord** records = malloc(record_capacity * sizeof(ClimateRecord*));

    FILE* file = fopen("climate-daily.csv", "r");
    char line[256];

    while (fgets(line, sizeof(line), file)) {
        // Resize the array when additional memory needed
        if (record_count == record_capacity) {
            record_capacity *= 2;
            records = realloc(records, record_capacity * sizeof(ClimateRecord*));
        }

        ClimateRecord* record = load_climate_data(line);
        if (record == NULL) {
            continue;
        }
        records[record_count++] = record;
    }

    fclose(file);

    analyzeData(records, record_count);
    userGeneratedReport(records, record_count);

    for (int i = 0; i < record_count; i++) {
        free(records[i]);
    }

    return 0;
}